picoCTF-web/api/logger.py

```diff
@@ -140,29 +139,47 @@ def setup_logs(args):
     stats_log.setLevel(logging.INFO)
     log.root.addHandler(stats_log)
 
+def _remove_parameter(arg_dict, param_path):
+    """Recurses through a dictionary of dictionaries and removes the given param"""
+    if param_path[0] not in arg_dict:
+        return arg_dict
 
-def log_action(f):
-    """Log a function invocation and its result."""
 
-    @wraps(f)
-    def wrapper(*args, **kwargs):
-        """Provide contextual information to the logger."""
-        log_information = {
-            "name": "{}.{}".format(f.__module__, f.__name__),
-            "args": args,
-            "kwargs": kwargs,
-            "result": None,
-        }
-        try:
-            log_information["result"] = f(*args, **kwargs)
-        except Exception as error:
-            log_information["exception"] = error
-            raise error
-        finally:
-            log.info(log_information)
-        return log_information["result"]
-
-    return wrapper
+    new_dict = arg_dict.copy() # Needed to avoid aliasing effects
+    if len(param_path) == 1:
+        new_dict[param_path[0]] = "REDACTED"
+        return new_dict
+
+    new_dict[param_path[0]] = _remove_parameter(arg_dict[param_path[0]], param_path[1:])
+    return new_dict
+
+def log_action(dont_log=[]):
+    """Log a function invocation and its result."""
+    def outer_wrapper(f):
+        @wraps(f)
+        def wrapper(*args, **kwargs):
+            """Provide contextual information to the logger."""
+
+            func_args = inspect.getcallargs(f, *args, **kwargs)
+            for param in dont_log:
+                param_path = param.split(".")
+                func_args = _remove_parameter(func_args, param_path)
+            log_information = {
+                "name": "{}.{}".format(f.__module__, f.__name__),
+                "args": func_args,
+                "result": None,
+            }
+            try:
+                log_information["result"] = f(*args, **kwargs)
+            except Exception as error:
+                log_information["exception"] = error
+                raise error
+            finally:
+                log.info(log_information)
+            return log_information["result"]
+
+        return wrapper
+    return outer_wrapper
```

Comment:
If log_action's signature is changed to:
```python
def log_action(_f=None, dont_log=[]):
```
and this line to:
```python
if _f is None:
    return outer_wrapper
else:
    return outer_wrapper(_f)
```
it would allow the decorator to be used with or without parens (if not setting anything for `dont_log`), and remove the need to change all existing empty `@log_action`s to the `@log_action()` form.